pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,269 @@
+				logger.Warningf("failed to check rgw health. %v", err)
+			}
+		}
+	}
+}
+
+func (c *rgwStatusChecker) checkObjectStoreHealth() error {
+	/*
+		0. purge everything by default
+		1. create an S3 user
+		2. always use the same user
+		3. if already exists just re-hydrate the s3 credentials
+		4. create a bucket with that user or use the existing one (always use the same bucket)
+		5. create a check file
+		6. get the MD5 of the file
+		7. PUT the file
+		8. GET the file
+		9. compare MD5
+		10. delete object on bucket
+		11. update CR health status check
+
+		Always keep the bucket and the user for the healthcheck, just do PUT and GET because bucket creation is expensive
+	*/
+
+	var s3AccessKey string
+	var s3SecretKey string
+	s3endpoint := fmt.Sprintf("%s:%s", BuildDomainName(c.objContext.Name, c.namespacedName.Namespace), c.port)
+
+	// Generate unique user and bucket name
+	bucket := genUniqueBucketName(c.objContext.UID)
+	userConfig := c.genUserConfig()
+
+	// Create S3 user
+	logger.Debugf("creating s3 user object %q", userConfig.UserID)
+	user, rgwerr, err := CreateUser(c.objContext, userConfig)
+	if err != nil {
+		if rgwerr == ErrorCodeFileExists {
+			objectUser, _, err := GetUser(c.objContext, userConfig.UserID)
+			if err != nil {
+				err = errors.Wrapf(err, "failed to get details from ceph object user %q", objectUser.UserID)
+				updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+				return err
+			}
+
+			// Set access and secret key
+			s3AccessKey = *objectUser.AccessKey
+			s3SecretKey = *objectUser.SecretKey
+
+		} else {
+			err = errors.Wrapf(err, "failed to create s3 user %q. error code %d", userConfig.UserID, rgwerr)
+			updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+			return err
+		}
+	} else {
+		// Set access and secret key
+		s3AccessKey = *user.AccessKey
+		s3SecretKey = *user.SecretKey
+	}
+
+	// Initiate s3 agent
+	logger.Debug("initializing s3 connection")
+	s3client := CreateNewS3Client(s3endpoint, s3AccessKey, s3SecretKey)
+
+	// Force purge before starting anything
+	err = cleanupObjectHealthCheck(s3client, c.objContext.UID)
+	if err != nil {
+		err = errors.Wrap(err, "failed to perform object cleanup")
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Purge on exit
+	defer cleanupObjectHealthCheck(s3client, c.objContext.UID)
+
+	// Create S3 bucket
+	logger.Debugf("creating bucket %q", bucket)
+	_, err = s3client.CreateBucket(bucket)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to create bucket %q", bucket)
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Put an object into the bucket
+	logger.Debugf("putting object %q in bucket %q", s3HealthCheckObjectKey, bucket)
+	_, err = s3client.PutObjectInBucket(bucket, string(s3HealthCheckObjectBody), s3HealthCheckObjectKey, contentType)
```

Comment:
Is there something in the PUT/GET that is unique to this cluster? If multiple clusters are connected to the same external cluster we wouldn't want the health check to conflict. Using the same user and bucket seems fine, but putting a different object sounds good.

Reply:
They can never get the same username and bucket since they are both prefixed by the CephObjectStore UID. So the object name does not matter if it's the same at this point.